### ISSUE TITLE
Support multiple Mesa versions?

### ISSUE THREAD

**Issue by Atemu** [NixOS/nixpkgs#189340]
## Issue description

You often need bleeding edge Mesa for gaming purposes and lesser platforms sometimes don't work with the newest Mesa versions.

What do you think of setting up a little infrastructure to support multiple versions of mesa that users can take bleeding-edge drivers from or can be used as the default on lesser platforms such as Darwin?

This could also have the advantage of shortening the feedback cycles for updates to mesa because, technically, the only patch that needs to go to through staging is the patch that sets the default mesa versions packages are compiled against.  
That way we could also have the newer mesa versions tested on master in nixosTests way before it's released to the masses via staging-next and, even more importantly, by users in actual real-world scenarios.

Precedent for this exists in the zfs and nvidia kernel modules.

cc @primeos @vcunat

**Comment by vcunat** [comment]
Well, I don't know why many packages need mesa as build input.  Most should be fine with libglvnd AFAIK.  I believe that's the first step to investigate.  By itself, reducing this would shorten the feedback cycle, regardless of actually supporting multiple versions.

**Comment by primeos** [comment]
@Atemu I'm not really sure how your approach would look like (i.e., how exactly would you propose implementing this?) but here are some notes from me:
- This would definitely require additional Mesa maintainers that commit to it (I'm already getting too slow so Mesa needs help as is). This is IMO the most important point to figure out. Someone needs to implement and maintain it.
- I'm not really that convinced that we would gain much. We could add a mesa-amber package for older hardware (e.g., Arch has one: https://archlinux.org/packages/extra/x86_64/mesa-amber/), Mesa should be quite backwards compatible, on nixpkgs-unstable we're already using the most recent Mesa version (apart from RCs and .0 development releases) so we'd basically have to switch to the `main` branch which is always undesirable for Nixpkgs IMO because how to decided how often to update and to which revisions? IMO an overlay would be much better for this purpose (i.e., no Nixpkgs changes should be required).
- It'd be mainly useful for nixos-stable (I guess someone could maintain something like a `mesa-rolling` package on the current stable release).
- Exchanging Mesa versions is currently unelegant (https://nixos.wiki/wiki/OpenGL#Testing_Mesa_updates) so this should ideally be improved first.

**Comment by joepie91** [comment]
> on nixpkgs-unstable we're already using the most recent Mesa version (apart from RCs and .0 development releases) 

I don't think that we are; the latest release for Mesa seems to be 22.1.5, but unstable currently ships 22.0.4.

**Comment by Atemu** [comment]
We ship 22.1.7 which is the newest stable version atm.

https://github.com/NixOS/nixpkgs/blob/fd745268a4f48d25875cc32b17abb0b32f45207e/pkgs/development/libraries/mesa/default.nix#L37
